## Decision Tree Classifier Note book



## importing libraries for this particular model

In [2]:

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


from sklearn.metrics import accuracy_score , classification_report 
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split


## Load dataset (importing dataset and looking at the lenght and breadth of the dataset)

In [3]:

df = pd.read_csv("C:\\Users\\HP\\Desktop\\etl\\railway.csv")
df.head()
df.shape

(31653, 18)

## checking out the characteristics of the data set

In [4]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Transaction ID       31653 non-null  object
 1   Date of Purchase     31653 non-null  object
 2   Time of Purchase     31653 non-null  object
 3   Purchase Type        31653 non-null  object
 4   Payment Method       31653 non-null  object
 5   Railcard             10735 non-null  object
 6   Ticket Class         31653 non-null  object
 7   Ticket Type          31653 non-null  object
 8   Price                31653 non-null  int64 
 9   Departure Station    31653 non-null  object
 10  Arrival Destination  31653 non-null  object
 11  Date of Journey      31653 non-null  object
 12  Departure Time       31653 non-null  object
 13  Arrival Time         31653 non-null  object
 14  Actual Arrival Time  29773 non-null  object
 15  Journey Status       31653 non-null  object
 16  Reas

## looking for columns with null values

In [5]:

df.isnull().sum()

Transaction ID             0
Date of Purchase           0
Time of Purchase           0
Purchase Type              0
Payment Method             0
Railcard               20918
Ticket Class               0
Ticket Type                0
Price                      0
Departure Station          0
Arrival Destination        0
Date of Journey            0
Departure Time             0
Arrival Time               0
Actual Arrival Time     1880
Journey Status             0
Reason for Delay       27481
Refund Request             0
dtype: int64

## analyzing the the dataset columns 

In [6]:

df.describe().T

count     mean        std  min  25%   50%   75%    max
Price  31653.0  23.4392  29.997628  1.0  5.0  11.0  35.0  267.0

## looking for spaces in the columns name 

In [7]:

df.columns

Index(['Transaction ID', 'Date of Purchase', 'Time of Purchase',
       'Purchase Type', 'Payment Method', 'Railcard', 'Ticket Class',
       'Ticket Type', 'Price', 'Departure Station', 'Arrival Destination',
       'Date of Journey', 'Departure Time', 'Arrival Time',
       'Actual Arrival Time', 'Journey Status', 'Reason for Delay',
       'Refund Request'],
      dtype='object')

## replacing underscores for spaces in the column name

In [8]:

df.columns = ['Transaction_ID', 'Date_of_Purchase', 'Time_of_Purchase',
       'Purchase_Type', 'Payment_Method', 'Railcard', 'Ticket_Class',
       'Ticket_Type', 'Price', 'Departure_Station', 'Arrival_Destination',
       'Date_of_Journey', 'Departure_Time', 'Arrival_Time',
       'Actual_Arrival_Time', 'Journey_Status', 'Reason_for_Delay',
       'Refund_Request']
df.columns

Index(['Transaction_ID', 'Date_of_Purchase', 'Time_of_Purchase',
       'Purchase_Type', 'Payment_Method', 'Railcard', 'Ticket_Class',
       'Ticket_Type', 'Price', 'Departure_Station', 'Arrival_Destination',
       'Date_of_Journey', 'Departure_Time', 'Arrival_Time',
       'Actual_Arrival_Time', 'Journey_Status', 'Reason_for_Delay',
       'Refund_Request'],
      dtype='object')

## Remove Duplicates

In [9]:

df.drop_duplicates('Transaction_ID')

Transaction_ID Date_of_Purchase Time_of_Purchase  \
0      da8a6ba8-b3dc-4677-b176        12/8/2023         12:41:11   
1      b0cdd1b0-f214-4197-be53       12/16/2023         11:23:01   
2      f3ba7a96-f713-40d9-9629       12/19/2023         19:51:27   
3      b2471f11-4fe7-4c87-8ab4       12/20/2023         23:00:36   
4      2be00b45-0762-485e-a7a3       12/27/2023         18:22:56   
...                        ...              ...              ...   
31648  1304623d-b8b7-4999-8e9c        4/30/2024         18:42:58   
31649  7da22246-f480-417c-bc2f        4/30/2024         18:46:10   
31650  add9debf-46c1-4c75-b52d        4/30/2024         18:56:41   
31651  b92b047c-21fd-4859-966a        4/30/2024         19:51:47   
31652  1d5d89a2-bde5-410f-8f91        4/30/2024         20:05:39   

      Purchase_Type Payment_Method Railcard Ticket_Class Ticket_Type  Price  \
0            Online    Contactless    Adult     Standard     Advance     43   
1           Station    Credit Card    Adult     Standard     Advance     23   
2            Online    Credit Card      NaN     Standard     Advance      3   
3           Station    Credit Card      NaN     Standard     Advance     13   
4            Online    Contactless      NaN     Standard     Advance     76   
...             ...            ...      ...          ...         ...    ...   
31648        Online    Credit Card      NaN     Standard    Off-Peak      4   
31649        Online    Contactless      NaN     Standard    Off-Peak     10   
31650       Station    Credit Card      NaN     Standard    Off-Peak      4   
31651       Station    Credit Card      NaN     Standard    Off-Peak     10   
31652       Station    Credit Card    Adult     Standard    Off-Peak      3   

           Departure_Station    Arrival_Destination Date_of_Journey  \
0          London Paddington  Liverpool Lime Street        1/1/2024   
1         London Kings Cross                   York        1/1/2024   
2      Liverpool Lime Street  Manchester Piccadilly        1/2/2024   
3          London Paddington                Reading        1/1/2024   
4      Liverpool Lime Street          London Euston        1/1/2024   
...                      ...                    ...             ...   
31648  Manchester Piccadilly  Liverpool Lime Street       4/30/2024   
31649          London Euston  Birmingham New Street       4/30/2024   
31650  Manchester Piccadilly  Liverpool Lime Street       4/30/2024   
31651          London Euston  Birmingham New Street       4/30/2024   
31652  Liverpool Lime Street  Manchester Piccadilly       4/30/2024   

      Departure_Time Arrival_Time Actual_Arrival_Time Journey_Status  \
0           11:00:00     13:30:00            13:30:00        On Time   
1            9:45:00     11:35:00            11:40:00        Delayed   
2           18:15:00     18:45:00            18:45:00        On Time   
3           21:30:00     22:30:00            22:30:00        On Time   
4           16:45:00     19:00:00            19:00:00        On Time   
...              ...          ...                 ...            ...   
31648       20:00:00     20:30:00            20:30:00        On Time   
31649       20:15:00     21:35:00            21:35:00        On Time   
31650       20:15:00     20:45:00            20:45:00        On Time   
31651       21:15:00     22:35:00            22:35:00        On Time   
31652       21:30:00     22:00:00            22:00:00        On Time   

      Reason_for_Delay Refund_Request  
0                  NaN             No  
1       Signal Failure             No  
2                  NaN             No  
3                  NaN             No  
4                  NaN             No  
...                ...            ...  
31648              NaN             No  
31649              NaN             No  
31650              NaN             No  
31651              NaN             No  
31652              NaN             No  

[31653 rows x 18 columns]

## checking the first five rows and the name of the column

In [10]:

df.head(5)

Transaction_ID Date_of_Purchase Time_of_Purchase Purchase_Type  \
0  da8a6ba8-b3dc-4677-b176        12/8/2023         12:41:11        Online   
1  b0cdd1b0-f214-4197-be53       12/16/2023         11:23:01       Station   
2  f3ba7a96-f713-40d9-9629       12/19/2023         19:51:27        Online   
3  b2471f11-4fe7-4c87-8ab4       12/20/2023         23:00:36       Station   
4  2be00b45-0762-485e-a7a3       12/27/2023         18:22:56        Online   

  Payment_Method Railcard Ticket_Class Ticket_Type  Price  \
0    Contactless    Adult     Standard     Advance     43   
1    Credit Card    Adult     Standard     Advance     23   
2    Credit Card      NaN     Standard     Advance      3   
3    Credit Card      NaN     Standard     Advance     13   
4    Contactless      NaN     Standard     Advance     76   

       Departure_Station    Arrival_Destination Date_of_Journey  \
0      London Paddington  Liverpool Lime Street        1/1/2024   
1     London Kings Cross                   York        1/1/2024   
2  Liverpool Lime Street  Manchester Piccadilly        1/2/2024   
3      London Paddington                Reading        1/1/2024   
4  Liverpool Lime Street          London Euston        1/1/2024   

  Departure_Time Arrival_Time Actual_Arrival_Time Journey_Status  \
0       11:00:00     13:30:00            13:30:00        On Time   
1        9:45:00     11:35:00            11:40:00        Delayed   
2       18:15:00     18:45:00            18:45:00        On Time   
3       21:30:00     22:30:00            22:30:00        On Time   
4       16:45:00     19:00:00            19:00:00        On Time   

  Reason_for_Delay Refund_Request  
0              NaN             No  
1   Signal Failure             No  
2              NaN             No  
3              NaN             No  
4              NaN             No

# **Data Preparation
    -

## replacing the columns name withe numeric figure for preparations of data modelling

In [12]:

df['ipurchase_type'] = pd.DataFrame(df.Purchase_Type.map({'Online' : 1 , 'Station' : 2 }))
df['iPayment_Method'] = pd.DataFrame(df.Payment_Method.map({'Contactless' : 1 , 'Credit Card' : 2 , 'Debit Card' : 3}))
df['iJourney_Status'] = pd.DataFrame(df.Journey_Status.map({'On Time' : 1 , 'Delayed' : 2 ,'Cancelled' : 3}))
df['iRefund_Request'] = pd.DataFrame(df.Refund_Request.map({'No' : 1 , 'Yes' : 2}))
df['iTicket_Class'] = pd.DataFrame(df.Ticket_Class.map({'Standard' :1 , 'First Class' : 2}))
df.head(5)
#df.shape

Transaction_ID Date_of_Purchase Time_of_Purchase Purchase_Type  \
0  da8a6ba8-b3dc-4677-b176        12/8/2023         12:41:11        Online   
1  b0cdd1b0-f214-4197-be53       12/16/2023         11:23:01       Station   
2  f3ba7a96-f713-40d9-9629       12/19/2023         19:51:27        Online   
3  b2471f11-4fe7-4c87-8ab4       12/20/2023         23:00:36       Station   
4  2be00b45-0762-485e-a7a3       12/27/2023         18:22:56        Online   

  Payment_Method Railcard Ticket_Class Ticket_Type  Price  \
0    Contactless    Adult     Standard     Advance     43   
1    Credit Card    Adult     Standard     Advance     23   
2    Credit Card      NaN     Standard     Advance      3   
3    Credit Card      NaN     Standard     Advance     13   
4    Contactless      NaN     Standard     Advance     76   

       Departure_Station  ... Arrival_Time Actual_Arrival_Time Journey_Status  \
0      London Paddington  ...     13:30:00            13:30:00        On Time   
1     London Kings Cross  ...     11:35:00            11:40:00        Delayed   
2  Liverpool Lime Street  ...     18:45:00            18:45:00        On Time   
3      London Paddington  ...     22:30:00            22:30:00        On Time   
4  Liverpool Lime Street  ...     19:00:00            19:00:00        On Time   

  Reason_for_Delay Refund_Request ipurchase_type iPayment_Method  \
0              NaN             No              1               1   
1   Signal Failure             No              2               2   
2              NaN             No              1               2   
3              NaN             No              2               2   
4              NaN             No              1               1   

  iJourney_Status  iRefund_Request  iTicket_Class  
0               1                1              1  
1               2                1              1  
2               1                1              1  
3               1                1              1  
4               1                1              1  

[5 rows x 23 columns]

# **Data preparation** 
### is a crucial step in machine learning and data modeling because raw data is often messy,
### inconsistent, and incomplete. Preparing data properly ensures better model performance,
### accuracy, and generalizability. Here are the key reasons why we prepare data:

## creating a dataset of numeric columns to be known as 'X' variable for preparation of data modelling

In [14]:

X =df.iloc[: , [8   , 18 , 19 ,20 ,21 , 22]]
#X.head()
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   Price            31653 non-null  int64
 1   ipurchase_type   31653 non-null  int64
 2   iPayment_Method  31653 non-null  int64
 3   iJourney_Status  31653 non-null  int64
 4   iRefund_Request  31653 non-null  int64
 5   iTicket_Class    31653 non-null  int64
dtypes: int64(6)
memory usage: 1.4 MB


## creating a column 'y' variable for preparation of data modelling

In [15]:

y = df.iloc[: ,[7]]
y.shape
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 1 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Ticket_Type  31653 non-null  object
dtypes: object(1)
memory usage: 247.4+ KB


## Split dataset into training and testing sets

In [16]:

X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.3, random_state=42)

## Initialize and train Decision Tree model

In [17]:

clf = DecisionTreeClassifier()
clf.fit(X_train , y_train)

DecisionTreeClassifier()

## the predicted column from the data model

In [18]:

df['predicted'] = y_pred = pd.DataFrame(clf.predict(X_test))
y_pred.shape

(9496, 1)

## comparison of the actual data vs the predicted data

In [19]:

L = df.iloc[ : ,[7 , 23]]
L.head(20)

Ticket_Type predicted
0      Advance   Advance
1      Advance  Off-Peak
2      Advance   Advance
3      Advance   Advance
4      Advance   Advance
5      Advance   Advance
6      Advance   Advance
7      Advance   Advance
8      Advance   Advance
9      Advance   Advance
10     Advance   Advance
11     Advance   Advance
12     Advance  Off-Peak
13     Advance  Off-Peak
14     Advance   Advance
15     Advance  Off-Peak
16     Advance   Advance
17     Advance   Advance
18     Advance   Advance
19     Advance   Advance

## Model evaluation

In [20]:

accuracy_score(y_test, y_pred)
#0.8073925863521483

0.8073925863521483

## Model evaluation

In [21]:

report = classification_report(y_test , y_pred )
print('Classification Report:\n', report )

Classification Report:
               precision    recall  f1-score   support

     Advance       0.77      0.95      0.85      5230
     Anytime       0.90      0.67      0.77      1631
    Off-Peak       0.87      0.62      0.72      2635

    accuracy                           0.81      9496
   macro avg       0.85      0.74      0.78      9496
weighted avg       0.82      0.81      0.80      9496



# **Metrics Explained**
## **Precision**
    -Measures how many of the predicted positive cases were actually correct
    -High precision means fewer false positives.

## **Recall** 
    -Measures how many of the actual positive cases were correctly predicted.
    -High recall means fewer false negatives.

## **F1-score** 

    -The harmonic mean of precision and recall.
    -A good balance between precision and recall, useful when you need both metrics to be high.

## **Support**

    -The number of actual occurrences of each class in the dataset.
    -Helps understand class distribution and whether the model is biased.

## **Summary of Findings from the Decision Tree Classifier Model**
## **1. Dataset Overview**
The dataset appears to be related to railway transactions, containing information on ticket purchases, payment methods, journey status, and refund requests.
Some preprocessing steps include:
Checking for missing values.
Removing duplicates based on Transaction_ID.
Renaming columns for better readability.
Converting categorical variables into numerical values for modeling.

## **2. Data Preprocessing**
Several categorical variables were mapped to numeric representations, such as:
Purchase_Type (Online = 1, Station = 2).
Payment_Method (Contactless = 1, Credit Card = 2, Debit Card = 3).
Journey_Status (On Time = 1, Delayed = 2, Cancelled = 3).
Refund_Request (No = 1, Yes = 2).
Ticket_Class (Standard = 1, First Class = 2).

## **3. Model Training**
Decision Tree Classifier was used to train the model.
The dataset was split into training and testing sets using train_test_split().
Features (X) and target variable (y) were defined appropriately.
4. Model Evaluation
The model's performance was evaluated using:
Accuracy Score: Measures overall correctness.
Classification Report: Provides precision, recall, and F1-score for each class.

# **Detailed Analysis of Model Performance**
## **Dataset Insights**
The dataset contains 31,653 records with 18 columns before preprocessing.
After feature selection, the dataset was reduced to 6 numerical features.
The target variable is likely related to ticket type, journey status, or refund request.
2. Model Performance
The extracted classification report includes:

# **Precision (how many predicted positives were actually positive)**
Recall (how many actual positives were correctly predicted)
F1-score (harmonic mean of precision and recall)
The visible part of the classification report suggests the model is predicting multiple classes.